src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -0,0 +1,221 @@
+                log.Error("Couldn't find type {GitExtTypeName}", gitExtTypeName);
+            }
+
+            var gitExt = serviceProvider.GetService(gitExtType);
+            if (gitExt == null)
+            {
+                log.Error("Couldn't find service for type {GitExtType}", gitExtType);
+                return;
+            }
+
+            dte = serviceProvider.TryGetService<DTE>();
+            if (dte == null)
+            {
+                log.Error("Couldn't find service for type {DteType}", typeof(DTE));
+            }
+
+            Refresh(gitExt);
+
+            var notifyPropertyChanged = gitExt as INotifyPropertyChanged;
+            if (notifyPropertyChanged == null)
+            {
+                log.Error("The service {ServiceObject} doesn't implement {Interface}", gitExt, typeof(INotifyPropertyChanged));
+                return;
+            }
+
+            notifyPropertyChanged.PropertyChanged += (s, e) => Refresh(gitExt);
+        }
+
+        /// <summary>
+        /// Used for unit testing.
+        /// </summary>
+        public interface IService
+        {
+            string FindTrackedSha(string repositoryPath);
+            ILocalRepositoryModel CreateRepository(string path);
+        }
+
+        class Service : IService
+        {
+            public string FindTrackedSha(string repositoryPath)
+            {
+                using (var repo = new Repository(repositoryPath))
+                {
+                    return repo.Head.TrackedBranch?.Tip.Sha;
+                }
+            }
+
+            public ILocalRepositoryModel CreateRepository(string path) => new LocalRepositoryModel(path);
+        }
+
+        void Refresh(object gitExt)
+        {
+            try
+            {
+                var repo = FindActiveRepository(gitExt);
+                var newSolutionPath = dte?.Solution?.FullName;
+
+                if (repo == null && newSolutionPath == solutionPath)
+                {
+                    // Ignore when ActiveRepositories is empty and solution hasn't changed.
+                    // https://github.com/github/VisualStudio/issues/1421
+                    log.Information("Ignoring no ActiveRepository when solution hasn't changed");
+                }
+                else
+                {
+                    var newRepositoryPath = repo?.RepositoryPath;
+                    var newBranchName = repo?.BranchName;
+                    var newHeadSha = repo?.HeadSha;
+                    var newTrackedSha = newRepositoryPath != null ? service.FindTrackedSha(newRepositoryPath) : null;
+
+                    if (newRepositoryPath != repositoryPath)
+                    {
+                        log.Information("Fire PropertyChanged event for ActiveRepository");
+                        ActiveRepository = repo != null ? service.CreateRepository(repo.RepositoryPath) : null;
+                        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ActiveRepository)));
```

Comment:
This should probably happen in the property setter directly, so that it's always fired regardless of where the property is set from (probably with some checks so it's not fired if the value is unchanged maybe?)

Reply:
Makes sense. Stick to a consistent pattern.